swarm/network/networkId_test.go

```diff
@@ -0,0 +1,271 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/state"
+)
+
+var (
+	vmodule          = flag.String("vmodule", "", "log filters for logger via Vmodule")
+	verbosity        = flag.Int("verbosity", 0, "log filters for logger via Vmodule")
+	currentNetworkId int
+	cnt              int
+	nodeMap          map[int][]discover.NodeID
+	kademlias        map[discover.NodeID]*Kademlia
+	quitC            chan struct{}
+	errc             chan error
+)
+
+const (
+	NumberOfNets = 4
+	MaxTimeout   = 6
+)
+
+func init() {
+	flag.Parse()
+	rand.Seed(time.Now().Unix())
+}
+
+/*
+Run the network ID test.
+The test creates one simulations.Network instance,
+a number of nodes, then connects nodes with each other in this network.
+
+Each node gets a network ID assigned according to the number of networks.
+Having more network IDs is just arbitrary in order to exclude
+false positives.
+
+Nodes should only connect with other nodes with the same network ID.
+After the setup phase, the test checks on each node if it has the
+expected node connections (excluding those not sharing the network ID).
+*/
+func TestNetworkId(t *testing.T) {
+	log.Debug("Start test")
+	//arbitrarily set the number of nodes. It could be any number
+	numNodes := 24
+	//the nodeMap maps all nodes (slice value) with the same network ID (key)
+	nodeMap = make(map[int][]discover.NodeID)
+	//set up the network and connect nodes
+	quitC = make(chan struct{})
+	errc = make(chan error)
+	net, err := setupNetwork(numNodes)
+	if err != nil {
+		t.Fatalf("Error setting up network: %v", err)
+	}
+	//channel to close disconnection watcher routine
+	defer func() {
+		//shutdown the snapshot network
+		log.Trace("Shutting down network")
+		net.Shutdown()
+	}()
+	//let's sleep to ensure all nodes are connected
+	//for each group sharing the same network ID...
+	for _, netIdGroup := range nodeMap {
+		log.Trace("netIdGroup size", "size", len(netIdGroup))
+		//...check that their size of the kademlia is of the expected size
```

Comment:
> Maybe it would be even safer to check for every node in the network whether all peers have the same network id?

Well the network ID gets assigned on node creation with the modulo. It shouldn't change (I hope there is some unit test for that....). The really safe check is to see if the nodes are in the kademlia table, as the test is doing after the size check.

The crucial thing here is the `Sleep` we need. I originally had a `Sleep` only, then moved on to implement a subscription as we try to avoid Sleeps. But the subscription is not enough yet, as it may still be building up connections. So it looks like for this test we need a `Sleep` to succeed.

Reply:
> The really safe check is to see if the nodes are in the kademlia table, as the test is doing after the size check.

Yes, my comment is to check for every node is all its connected peers have the same network id. It should not be possible for a peer to change network id, but a node can check if all peers in kademlia have the same network id.

Regarding the sleep, I am wondering is 4s too much? I have all passes for 10ms (probably too short for Travis), for 5ms it fails very often.

Reply:
Yes 4s is a lot. I was just wanting to go the safe side. If you played around and had good results with 10ms, probably we can use 1s?

Reply:
👍 1s is a good safe value.